[{"count": 0, "tags": [], "text": "Created attachment 35564\njstack thread dump\n\nWe are observing a scenario when NIO2 connector on tomcat becomes unresponsive after some period of time, at the same time NIO connector running on the same host is still able to process the same requests and serves traffic. Only server restart helps in this case.\nThis issue is intermittent and with the current infrastructure we have few nodes behind LB and it happens from time to time (like once per week) for each node, so it seems to be not a node or hardware specific in our case.\nBelow is our server.xml:\n\n    <Executor name=\"tomcatThreadPool\" namePrefix=\"catalina-exec-\" maxThreads=\"800\" minSpareThreads=\"100\"/>\n\n\n      <Connector executor=\"tomcatServiceThreadPool\"\n                 port=\"8080\"\n                 protocol=\"org.apache.coyote.http11.Http11Nio2Protocol\"\n                 connectionTimeout=\"1000\"\n                 enableLookups=\"false\"\n                 acceptorThreadCount=\"1\"\n                 processorCache=\"800\"\n                 socket.tcpNoDelay=\"true\"\n                 socket.soKeepAlive=\"true\"\n                 socket.soLingerOn=\"false\"\n                 compression=\"256\"\n                 compressableMimeType=\"text/html,text/xml,text/plain,application/x-protobuf,application/json,application/javascript\"\n                 URIEncoding=\"UTF-8\" />\n\n      <!-- The load balancer terminates SSL connections and\n           then forwards them to the following connector as\n           normal HTTP (non-secure) requests\n       -->\n      <Connector executor=\"tomcatServiceThreadPool\"\n                 port=\"8443\"\n                 protocol=\"org.apache.coyote.http11.Http11NioProtocol\"\n                 connectionTimeout=\"1000\"\n                 enableLookups=\"false\"\n                 connectionLinger=\"-1\"\n                 acceptorThreadCount=\"20\"\n                 processorCache=\"800\"\n                 socket.tcpNoDelay=\"true\"\n                 socket.soKeepAlive=\"true\"\n                 socket.soLingerOn=\"false\"\n                 compression=\"256\"\n                 compressableMimeType=\"text/html,text/xml,text/plain,application/x-protobuf,application/json,application/javascript\"\n                 URIEncoding=\"UTF-8\" />\n\n\n      <!-- Define an AJP 1.3 Connector on port 8009 -->\n      <Connector port=\"8009\" protocol=\"AJP/1.3\" redirectPort=\"8443\" />\n\nAlso below is an example of the behavior we observe:\n\ncurl -verbose 'http://localhost:8080/rs?id=nio2issue'\n* About to connect() to localhost port 8080 (#0)\n*   Trying 127.0.0.1... connected\n* Connected to localhost (127.0.0.1) port 8080 (#0)\n> GET /rs?id=nio2issue HTTP/1.1\n> User-Agent: curl/7.19.7 (x86_64-redhat-linux-gnu) libcurl/7.19.7 NSS/3.15.3 zlib/1.2.3 libidn/1.18 libssh2/1.4.2\n> Host: localhost:8080\n> Accept: */*\n> Referer: rbose\n>\n* Closing connection #0\n* Failure when receiving data from the peer\ncurl: (56) Failure when receiving data from the peer\n\nat the same time:\n\ncurl -i 'http://localhost:8443/rs?id=nio2issue'\nHTTP/1.1 302 Found\n\nAlso, no unusual errors are logged to catalina.out at the time of the accident. Enclosed is thread dump from the server.\nAlso, we have observed the same behavior on tomcat 8.0.18 and upgraded to the latest version in the same release 8.0.47 but it didn't help.\n\nPlease let me know what else might be helpful as we keep one of the servers in this state, for now, to be able to gather any data as the issue is intermittent and we were not able to reproduce with a simple load test.\n\nRegards,\nOleg.", "is_private": false, "bug_id": 61831, "id": 202460, "time": "2017-11-29T01:58:29Z", "creator": "yaolega@gmail.com", "creation_time": "2017-11-29T01:58:29Z", "attachment_id": 35564}, {"count": 1, "tags": [], "bug_id": 61831, "text": "The thread dump looks perfect: acceptor thread blocking on the accept, all threads idle and ready to execute something. Please investigate on the user list to get at least some idea on how to reproduce it.\nIf possible, try to avoid using a custom executor, it makes things more complex and the benefit is usually not obvious.", "id": 202463, "time": "2017-11-29T06:58:21Z", "creator": "remm@apache.org", "creation_time": "2017-11-29T06:58:21Z", "is_private": false, "attachment_id": null}, {"count": 2, "tags": [], "text": "Hi, \n\nI realize that thread dump might look fine and this is the most confusing part: even simple curl command from the same host receives no response from this connector and it starts working fine after tomcat restart.  At the same time tomcat in overall looks to be healthy and another connector works fine, as this happens from time to time on different servers, this doesn't look like to be OS or hardware issue but something which is tomcat NIO2 specific. \nAnd when we do any request to this NIO2 endpoint connector in a bad state - no thread is triggered in tomcat, while looking into tomcat source code it looks like that countdownlatch was not simply updated and service just hangs because of this but the root cause is still not clear. \n\nSo curious what additional information we can provide to help investigate this issue together with tomcat apache dev team? \n\n\nAlso, I'm not sure about your remark about custom executor - we don't use custom one, we just configure the one form tomcat.\n\nRegards,\nOleg.", "attachment_id": null, "bug_id": 61831, "id": 202482, "time": "2017-11-29T19:25:56Z", "creator": "yaolega@gmail.com", "creation_time": "2017-11-29T19:25:56Z", "is_private": false}, {"count": 3, "tags": [], "bug_id": 61831, "attachment_id": null, "id": 202700, "time": "2017-12-11T11:52:38Z", "creator": "piotrek.kowalski@gmail.com", "creation_time": "2017-12-11T11:52:38Z", "is_private": false, "text": "We think we figured it out to be a Java Bug in asynchronous server socket implementation.\n\nPlease see the following bug report which seems to exhibit a similar issue.\nhttps://bugs.openjdk.java.net/browse/JDK-8172750"}, {"count": 4, "tags": [], "bug_id": 61831, "text": "Ok, maybe. Let us know if you find some elements demonstrating an issue in Tomcat.", "id": 202703, "time": "2017-12-11T12:37:03Z", "creator": "remm@apache.org", "creation_time": "2017-12-11T12:37:03Z", "is_private": false, "attachment_id": null}]